src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +87,75 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
-        /// The API streams results back to the caller.
+        /// Invokes API on the callback object hosted in the original process (usually devenv) associated with the
+        /// currently executing brokered service hosted in ServiceHub process. The API streams results back to the
+        /// caller.
         /// </summary>
-        /// <inheritdoc cref="BrokeredServiceConnection{TService}.InvokeStreamingServiceAsync"/>
+        /// <param name="invocation">A callback to asynchronously write data. The callback should not complete the <see
+        /// cref="PipeWriter"/>, but no harm will happen if it does.</param>
+        /// <param name="reader">A callback to asynchronously read data. The callback should not complete the <see
+        /// cref="PipeReader"/>, but no harm will happen if it does.</param>
+        /// <param name="cancellationToken">A cancellation token the operation will observe.</param>
         public async ValueTask<TResult> InvokeAsync<TResult>(
             Func<T, PipeWriter, CancellationToken, ValueTask> invocation,
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
             try
             {
-                return await BrokeredServiceConnection<T>.InvokeStreamingServiceAsync(_callback, invocation, reader, cancellationToken).ConfigureAwait(false);
+                return await InvokeWorkerAsync(_callback).ConfigureAwait(false);
             }
             catch (Exception exception) when (ReportUnexpectedException(exception, cancellationToken))
             {
                 throw OnUnexpectedException(exception, cancellationToken);
             }
+
+            async ValueTask<TResult> InvokeWorkerAsync(T service)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var pipe = new Pipe();
+
+                // Kick off the work to do the writing to the pipe in a fire-and-forget fashion.  It will start hot and
+                // will be able to do work as the reading side attempts to pull in the data it is writing.
+
+                _ = WriteAsync(service, pipe.Writer);
+
+                try
+                {
+                    return await reader(pipe.Reader, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // ensure we always complete the reader so the pipe can clean up all its resources. in the case of
+                    // an exception, attempt to complete the reader with that as well as that will tear down the writer
+                    // allowing it to stop writing and allowing the pipe to be cleaned up.
+                    await pipe.Reader.CompleteAsync(e).ConfigureAwait(false);
+                    throw;
```

Comment:
i believe this 'throw' is def necessary to ensure that either an exception on reading (or one on writing that is passed to hte reader) properly gets populated to hte caller of InvokeAsync